File: fortran/non-lineal-structures/btree/btree.f90
! BTree Knuth Implementation
! Luis Espino 2024

program BTree
    	implicit none

      	! Order 4
    	integer, parameter :: MAXI = 3, MINI = 1 

    	type nodeptr
        	type (BTreeNode), pointer :: ptr => null()
    	end type nodeptr
    
    	type BTreeNode
        	integer :: val(0:MAXI+1)
        	integer :: num = 0
        	type(nodeptr) :: link(0:MAXI+1)
    	end type BTreeNode

    	type(BTreeNode), pointer :: root => null()
    	
    	call insert(5)
    	call insert(10)
    	call insert(15)
    	call insert(20)
    	call insert(25)
    	call insert(30)
    	call insert(35)
    	call insert(40)
    	call insert(45)
    	call traversal(root)
    	print * , ''
    	
contains

subroutine insert(val)
        integer, intent(in) :: val
        integer :: i
        type(BTreeNode), pointer :: child
        allocate(child)
        if (setValue(val, i, root, child)) then
            	root => createNode(i, child)
        end if
end subroutine insert

recursive function setValue(val, pval, node, child) result(res)
        integer, intent(in) :: val
        integer, intent(inout) :: pval
        type(BTreeNode), pointer, intent(inout) :: node
        type(BTreeNode), pointer, intent(inout) :: child
        type(BTreeNode), pointer :: newnode        
        integer :: pos
        logical :: res
        allocate(newnode)
        if (.not. associated(node)) then            
            	pval = val
            	child => null()
            	res = .true.
            	return
        end if
        if (val < node%val(1)) then
            	pos = 0
        else
            	pos = node%num
            	do while (val < node%val(pos) .and. pos > 1) 
        		pos = pos - 1
            	end do
            	if (val == node%val(pos)) then
                	print *, "Duplicates are not permitted"
                	res = .false.
                	return
            	end if
        end if
        if (setValue(val, pval, node%link(pos)%ptr, child)) then
            	if (node%num < MAXI) then
                	call insertNode(pval, pos, node, child)
            	else
                	call splitNode(pval, pval, pos, node, child, newnode)
                	child => newnode
                	res = .true.
                return
            end if
        end if
        res = .false.
end function setValue

subroutine insertNode(val, pos, node, child)
        integer, intent(in) :: val, pos
        type(BTreeNode), pointer, intent(inout) :: node
        type(BTreeNode), pointer, intent(in) :: child
        integer :: j
        j = node%num
        do while (j > pos)
            	node%val(j + 1) = node%val(j)
            	node%link(j + 1)%ptr => node%link(j)%ptr
            	j = j - 1
        end do
        node%val(j + 1) = val
        node%link(j + 1)%ptr => child
        node%num = node%num + 1
    end subroutine insertNode

subroutine splitNode(val, pval, pos, node, child, newnode)
        integer, intent(in) :: val, pos
        integer, intent(inout) :: pval
        type(BTreeNode), pointer, intent(inout) :: node,  newnode
        type(BTreeNode), pointer, intent(in) ::  child
        integer :: median, i, j
        if (pos > MINI) then
            	median = MINI + 1
        else
            	median = MINI
        end if
        if (.not. associated(newnode)) then
        	allocate(newnode)
		do i = 0, MAXI
            		newnode%link(i)%ptr => null()
        	enddo
        end if
        j = median + 1
        do while (j <= MAXI)
            	newnode%val(j - median) = node%val(j)
            	newnode%link(j - median)%ptr => node%link(j)%ptr
            	j = j + 1
        end do
        node%num = median
        newnode%num = MAXI - median
        if (pos <= MINI) then
            	call insertNode(val, pos, node, child)
        else
            	call insertNode(val, pos - median, newnode, child)
        end if        
        pval = node%val(node%num)        
        newnode%link(0)%ptr => node%link(node%num)%ptr
        node%num = node%num - 1
end subroutine splitNode

function createNode(val, child) result(newNode)
        integer, intent(in) :: val
        type(BTreeNode), pointer, intent(in) :: child
        type(BTreeNode), pointer :: newNode
        integer :: i
        allocate(newNode)
        newNode%val(1) = val
        newNode%num = 1
        newNode%link(0)%ptr => root
        newNode%link(1)%ptr => child
        do i = 2, MAXI
            	newNode%link(i)%ptr => null()
        end do
end function createNode

recursive subroutine traversal(myNode)
        type(BTreeNode), pointer, intent(in) :: myNode
        integer :: i
        if (associated(myNode)) then
            	write (*, '(A)', advance='no') ' [ '
            	i = 0
            	do while (i < myNode%num)
                	write (*,'(1I3)', advance='no') myNode%val(i+1)
                	i = i + 1
            	end do
            	do i = 0, myNode%num
                	call traversal(myNode%link(i)%ptr)    
            	end do
            	write (*, '(A)', advance='no') ' ] '
        end if
end subroutine traversal

end program BTree

File: cpp/non-linear-structures/btree.cpp
// BTree Knuth Implementation
// Luis Espino 2024

#include <iostream>

// order 4
#define MAX 3
#define MIN 1

using namespace std;

class BTreeNode
{
    public:
    int val[MAX + 1];
    int num = 0;
    BTreeNode *link[MAX + 1];
};

class BTree
{
    public:
    BTreeNode *root = nullptr;

    void insert(int val)
    {
        int i;
        BTreeNode *child = nullptr;
        if (setValue(val, &i, root, &child))
            root = createNode(i, child);
    }

    int setValue(int val, int *pval, 
        BTreeNode *node, BTreeNode **child)
    {
        int pos;
        if (!node)
        {
            *pval = val;
            *child = nullptr;
            return 1;
        }

        if (val < node->val[1]) 
        {
            pos = 0;
        }
        else
        {
            for (pos = node->num;
                (val < node->val[pos] && pos > 1); pos--);
            if (val == node->val[pos]) 
            {
                cout << "Duplicates not allowed" << endl;
                return 0;
            }
        }
        if (setValue(val, pval, node->link[pos], child))
        {
            if (node->num < MAX)
            {
                insertNode(*pval, pos, node, *child);
            } 
            else 
            {
                splitNode(*pval, pval, pos, node, *child, child);
                return 1;
            }
        }
        return 0;
    }

    BTreeNode* createNode(int val, BTreeNode *child)
    {
        BTreeNode *newNode = new BTreeNode();
        newNode->val[1] = val;
        newNode->num = 1;
        newNode->link[0] = root;
        newNode->link[1] = child;
        return newNode;
    }

    void insertNode(int val, int pos, 
        BTreeNode *node, BTreeNode *child)
    {
        int j = node->num;
        while (j > pos) {
            node->val[j + 1] = node->val[j];
            node->link[j + 1] = node->link[j];
            j--;
        }
        node->val[j + 1] = val;
        node->link[j + 1] = child;
        node->num++;
    }

    void splitNode(int val, int *pval, int pos, 
        BTreeNode *node, BTreeNode *child, BTreeNode **newNode)
    {
        int median, j;
        if (pos > MIN)
            median = MIN + 1;
        else
            median = MIN;
        *newNode = new BTreeNode();
        j = median + 1;
        while (j <= MAX)
        {
            (*newNode)->val[j - median] = node->val[j];
            (*newNode)->link[j - median] = node->link[j];
            j++;
        }
        node->num = median;
        (*newNode)->num = MAX - median;

        if (pos <= MIN)
        {
            insertNode(val, pos, node, child);
        } 
        else
        {
            insertNode(val, pos - median, *newNode, child);
        }
        *pval = node->val[node->num];
        (*newNode)->link[0] = node->link[node->num];
        node->num--;
    }

    void traversal(BTreeNode *myNode)
    {
        int i;
        if (myNode)
        {
            cout << "[";
            for (i = 0; i < myNode->num; i++)
            {
                //traversal(myNode->link[i]);
                cout << myNode->val[i + 1] << ",";
            }
            for (i = 0; i <= myNode->num; i++)
            {
                traversal(myNode->link[i]);
            }
            cout << "]";
        }
    }

};

int main()
{
    BTree *t = new BTree();
    t->insert(5);t->insert(10);
    t->insert(15);t->insert(20);
    t->insert(25);t->insert(30);
    t->insert(35);t->insert(40);
    t->insert(45);t->insert(50);
    t->traversal(t->root);cout << endl;
}

